slatedb/src/clock.rs

```diff
@@ -0,0 +1,224 @@
+use std::{
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc,
+    },
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use tracing::info;
+
+use crate::{config::DurabilityLevel, SlateDBError};
+
+/// defines the clock that SlateDB will use during this session
+pub trait Clock: Debug + Send + Sync {
+    /// Returns a timestamp (typically measured in millis since the unix epoch),
+    /// must return monotonically increasing numbers (this is enforced
+    /// at runtime and will panic if the invariant is broken).
+    ///
+    /// Note that this clock does not need to return a number that
+    /// represents the unix timestamp; the only requirement is that
+    /// it represents a sequence that can attribute a logical ordering
+    /// to actions on the database.
+    fn now(&self) -> i64;
+
+    /// Returns the current time as a SystemTime.
+    ///
+    /// This function panics if the Clock time cannot be converted to a SystemTime.
+    fn now_systime(&self) -> SystemTime {
```

Comment:
if we end up splitting it into two traits it makes sense to only have this method in the SystemClock trait.